### ISSUE TITLE
Chef Server 12 does not allow client to be recrerated

### ISSUE THREAD

**Issue by tokenrain** [opscode/chef-server#54]
We rebuild many of our hosts regularly.

In the past we simply run a 'knife client delete <name>' and let the validator recreate the client upon bootstrap. We leave the node intact and it converges as per it's saved run_list.

Under chef 12 when we try and do the same thing we get

[2015-01-04T13:51:00-08:00] ERROR: 403 "Forbidden"
[2015-01-04T13:51:00-08:00] FATAL: Chef::Exceptions::ChildConvergeError: Chef run process exited unsuccessfully (exit code 1)

It appears that the server does not like the new client key in the sense that it is not associated with the old node.

This seems like a function that should work and would be useful for a large number of Chef users.

I am running Chef Server 12.0.1 on Centos 6.6

**Comment by mmzyk** [comment]
I assume you've upgraded from open source Chef Server 11 @tokenrain? I believe this has to do with the fuller permission systems in Chef Server 12. I think there was a thread about this on the mailing list that I will try to find, but I believe this is not a bug, but might require you to change a bit how you've been doing things. Let me look into it some.

**Comment by mmzyk** [comment]
@tokenrain Here's the mailing list thread I was thinking about: is this what you are encountering?

http://lists.opscode.com/sympa/arc/chef/2015-01/msg00037.html

**Comment by tokenrain** [comment]
This was actually me posting this issue on the Chef Mailing list. Basically in OpenSource Chef 11 server you could delete a client "knife client delete xxx" and if the client had a validator when it ran chef-client it could generate a new client key-pair which has the ability to write the existing node data because the name of the new client and the old node were the same. This made rebuilding hosts really simple as all you needed to do was delete the client and run chef-client when the host came back up. You did not have to save the environment and run list and pass that to the initial run.

I installed Chef 12 Server and found that you can no longer delete the client and create a new one and have the new client have write access to the old node data. This means that in order to rebuild a host I know have to obtain the old environment and run list, delete the node and the client, and supply the environment and run_list during the initial chef run.

My question for Chef Inc is simply is this a feature or a bug of Chef 12 Server. If it is a bug and is going to be fixed that is awesome and I will wait for the fix before moving everything over to Chef Server 12. If it is a feature then I will have to build a bit of code to automate adding an environment and run_list on rebuild.

Noah seemed to think that this was caused by inconsistent maybe buggy code but I did not get the feeling (Nor would I ask him to) that he knew exactly which code was causing the problem.

So TL;DR - Is this a bug or a feature

**Comment by stevendanna** [comment]
@tokenrain This is a feature. It is the expected outcome of the new ACL system.  The fact that it broke your workflow, however, is definitely something we typically hope to avoid and perhaps points to an improvement we could make in the future.  In the near term, I would recommend building the bit of code needed to rebuild nodes given this behavior.  I (and others) can certainly help you do that.

In Chef 11, most permissions decisions where made on the basis of whether you were an admin or non-admin.  However, a special exception was made if you were a client trying to operate on a node of the same name.  In this case, we allowed you to update the node.  Thus, client "foobar" could update node "foobar" even though they were not an admin client.

In Chef 12, only clients with the UPDATE permission on the node can update the node.  When a client initially creates a node, it gets UPDATE (as well as all other permissions on that node).  When you delete that client and recreate it, from the perspective of the ACL system it is a brand new client with no permissions on that node.  Thus, you either need to delete both the client and node (so that the new client will be the first client to create the node on the next chef-client run) and devise a way to save the run_list.  (You might also be able to come up with a process where you use the ACL API to fix the client permissions rather than delete the node, but either way you'll be writing some new custom code).

**Comment by tokenrain** [comment]
Thanks for the timely response and the excellent explanation to the problem. I also appreciate the offer of  help with building the code that is needed to deal with this, but I can handle that. Thanks!

**Comment by isavcic** [comment]
Sorry for necroing this thread, but is there a way to override this illogical and unnecessary behavior, at least from my point of view and compared to previous Chef versions? Thanks.

**Comment by spajus** [comment]
@isavcic you can find a script which fixed the permissions in https://docs.chef.io/auth.html (Client Key Permissions section). But I also agree that this feature causes a lot of frustration. Migrating chef server to different box is a painful process.

**Comment by isavcic** [comment]
@spajus Yeah, I get that, and that part was easy, when we migrated Chef server to a new version. We've used knife-acl, looping over clients and giving access to their node "counterparts"...

What's really frustrating is my understanding of how the new procedure of deleting and re-registering single clients now looks like:

1. delete the client
2. do a chef-client run on the client just to re-register it
3. change ACL so the client has access to the node of the same name
4. run chef-client again, so it's run can actually complete

Did I miss something? Is there a more logical and less painful way of doing this?

**Comment by Tech356** [comment]
@isavcic I solved this by running `client reregister` commands then putting the new cert in place. That way the client is never deleted so the permissions are not broken. Unfortunately this requires an administrator level service account to do the reregister.

**Comment by isavcic** [comment]
@Tech356 I see, but in that case you have to copy the newly generated client key by hand, or to assume you have the ssh pubkey auth in place so you can automate it to some extent...

**Comment by pepov** [comment]
We are not really happy with it, but solved the issue during the upgrade by adding group level access on the nodes container to the clients. However we are not sure that it won't cause us trouble in the future, perhaps with the next upgrade and would like to know whether you have something work in progress for this @stevendanna ?

**Comment by isavcic** [comment]
Yeah @stevendanna, should we brace for more breaking changes? Thanks in advance for the heads up!

**Comment by stevendanna** [comment]
@isavcic @pepov Unfortunately there is no way to revert to the current behavior.  As far as I know, we don't have upcoming changes that would break the work arounds you have outlined.  

Additionally, regarding further breaking changes, we've recently introduced an API versioning scheme that allows clients to signal which version of the API they are expecting.  Our intent is for future breaking changes to be versioned in this way so that we can make improvements to the API more rapidly while still supporting older clients.  

We do have a few upcoming, minor policy changes for ACLs (related to the default permissions we ship with and related to preventing users from locking themselves out of their organizations); I doubt we would bump the API version for these, since as far as we know they shouldn't break any existing workflows.

I hope this helps.  Our apologies that the new authorization system broke your previous workflows.  Please let us know if you need any help integrating with the new system.

**Comment by joerocklin** [comment]
@stevendanna Are the upcoming changes documented anywhere so we can look now and see if they might break things? 

It's hard to say definitively if an API change will break workflows since the API is open. To be safe: if there's a change to an API call in a released product, it should require a version bump.

**Comment by stevendanna** [comment]
@joerocklin The policy changes I was refering to are:

https://github.com/chef/chef-server/commit/a6fc2c38cf22acbbdfd49e6ebb7fafd6cd32208e
https://github.com/chef/chef-server/pull/248/files

The first allows users who share an org to read each other's public keys.  There will be a documentation update and announcement about this change prior to the release.   

The second is still under discussion but would prevent ACL updates that remove the admin group from the GRANT ace of an object.  The goal is to prevent users from getting objects into an unrepairable state.

**Comment by mmzyk** [comment]
It should be noted that https://github.com/chef/chef-server/pull/248 was just merged to master and should ship in the 12.2 server.
